text/0000-rustdoc-json.md

```diff
@@ -0,0 +1,1124 @@
+of the form:
+
+```rust
+type Foo<T> = Clone + std::fmt::Debug + Into<T>;
+```
+
+Name          | Type     | Description
+--------------|----------|-------------------------------------------------------------------------
+`bounds`      | [[GenericBound](#GenericBound)] | The trait bounds on the right hand side.
+`generics`    | [Generics](#Generics) | Any generic parameters on the type itself.
+
+### `kind == "foreign_type"`
+
+`inner` contains no fields. This item represents a type declaration in an extern block (see [here](https://github.com/rust-lang/rfcs/blob/master/text/1861-extern-types.md)
+for more details):
+
+```rust
+extern {
+    type Foo;
+}
+```
+
+### `kind == "extern_crate"`
+
+Name     | Type     | Description
+---------|----------|-------------------------------------------------------------------------
+`name`   | String   | The name of the extern crate.
+`rename` | String   | (*Optional*) The renaming of this crate with `extern crate foo as bar`.
+
+### `kind == "import"`
+
+Name     | Type     | Description
+---------|----------|-------------------------------------------------------------------------
+`source` | String   | The full path being imported (e.g. `"super::some_mod::other_mod::Struct"`).
+`name`   | String   | The name of the imported item (may be different from the last segment of `source` due to import renaming: `use source as name`).
+`id`     | [ID](#ID) | (*Optional*) The ID of the item being imported.
+`glob`   | bool     | Whether this import ends in a glob: `use source::*`.
+
+### `kind == "macro"`
+
+A `macro_rules!` declarative macro. Contains a single string with the source representation of
+the macro with the patterns stripped, for example:
+
+```rust
+macro_rules! vec {
+    () => { ... };
+    ($elem:expr; $n:expr) => { ... };
+    ($($x:expr),+ $(,)?) => { ... };
+}
+```
+
+TODO: proc macros
+
+## Span
+
+Name       | Type     | Description
+-----------|----------|----------------------------------------------------------------------------
+`filename` | String   | The path to the source file for this span relative to the crate root.
+`begin`    | (int, int) | The zero indexed line and column of the first character in this span.
+`begin`    | (int, int) | The zero indexed line and column of the last character in this span.
+
+## Deprecation
+
+Name       | Type     | Description
+-----------|----------|----------------------------------------------------------------------------
+`since`    | String   | (*Optional*) Usually a version number when this Item first became deprecated.
+`note`     | String   | (*Optional*) The reason for deprecation and/or what alternatives to use.
+
+## FnDecl
+
+Name       | Type     | Description
+-----------|----------|----------------------------------------------------------------------------
+`inputs`   | [(String, [Type](#Type))] | A list of parameter names and their types.
```

Comment:
This is not sufficient to describe function signatures in Rust - you could have arbitrary patterns.

```rust
fn f((_, x): (usize, usize)) { ... }
fn g(MyStruct { field1: usize, .. }) { ... }
```

Reply:
Woah didn't know you could destructure stuff like that in function args. I think it's still fine though, patterns just get stringified like names so you end up with `["(_, x)", <Type representation of tuple of usizes>]` and `["MyStruct { field1: usize, .. }", <Type representation of MyStruct>]` for those examples. It would be a problem if the consumer of the json wanted to know the names `x` and `field1` because they'd have to parse arbitrary patterns, but I can't think of a case where they need that info off the top of my head.

Reply:
I'm ok with having that in the output, but it should be marked as unstable and subject to change.

Reply:
+1 to marking the format of the value as unstable and subject to change, but I think we could stabilize its _existence_ and that it's a reasonable thing to put in the docs. (Here and elsewhere)

Reply:
Yeah, that seems reasonable. "You can display this to the end-user but you can't depend on parsing it".